solidity/contracts/bridge/Bridge.sol

```diff
@@ -795,10 +935,500 @@ contract Bridge is Ownable {
+    ///      - `mainUtxo` components must point to the recent main UTXO
+    ///        of the given wallet, as currently known on the Ethereum chain.
+    ///      - `redeemerOutputHash` must be proper 20-byte or 32-byte BTC hash
+    ///      - `redeemerOutputHash` cannot be the same as `walletPubKeyHash`
+    ///      - `amount` must be above or equal the `redemptionDustThreshold`
+    ///      - Given `walletPubKeyHash` and `redeemerOutputHash` pair can be
+    ///        used for only one pending request at the same time
+    ///      - Wallet must have enough Bitcoin balance to proceed the request
+    ///      - Redeemer must make an allowance in the Bank that the Bridge
+    ///        contract can spend the given `amount`.
+    function requestRedemption(
+        bytes20 walletPubKeyHash,
+        BitcoinTx.UTXO calldata mainUtxo,
+        bytes calldata redeemerOutputHash,
+        uint64 amount
+    ) external {
+        require(!faultyWallets[walletPubKeyHash], "Wallet marked as faulty");
+
+        bytes32 mainUtxoHash = mainUtxos[walletPubKeyHash];
+        require(mainUtxoHash != bytes32(0), "No main UTXO for given wallet");
+        require(
+            keccak256(
+                abi.encodePacked(
+                    mainUtxo.txHash,
+                    mainUtxo.txOutputIndex,
+                    mainUtxo.txOutputValue
+                )
+            ) == mainUtxoHash,
+            "Invalid main UTXO data"
+        );
+
+        // TODO: Check if `walletPubKeyHash` should be validated in some other ways.
+
+        // Validate if redeemer output hash has a correct length corresponding
+        // to a valid Bitcoin hash. P2PKH, P2WPKH and P2SH outputs will
+        // have 20-byte hashes. P2WSH outputs will have 32-byte hash.
+        require(
+            redeemerOutputHash.length == 20 || redeemerOutputHash.length == 32,
+            "Incorrect redeemer output hash length"
+        );
+
+        require(
+            keccak256(abi.encodePacked(walletPubKeyHash)) !=
+                keccak256(redeemerOutputHash),
+            "Redeemer output hash must not be equal to the wallet PKH"
+        );
+
+        require(
+            amount >= redemptionDustThreshold,
+            "Redemption amount too small"
+        );
+
+        // The redemption key is built on top of wallet public key hash
+        // and redeemer output hash pair. That means there can be only one
+        // request asking for redemption from specific wallet to the given
+        // BTC hash in the same time.
+        uint256 redemptionKey =
+            uint256(
+                keccak256(
+                    abi.encodePacked(walletPubKeyHash, redeemerOutputHash)
+                )
+            );
+
+        require(
+            // slither-disable-next-line incorrect-equality
+            pendingRedemptionRequests[redemptionKey].requestedAt == 0,
+            "Pending request with same redemption key already exists"
```

Comment:
Maybe something more user-friendly like:

```suggestion
            "There is a pending redemption request from this wallet to the same address"
```

The current version makes total sense to me but I am afraid it's not something users would understand when looking at JS console of their browser.

Reply:
See https://github.com/keep-network/tbtc-v2/pull/128/commits/bdcf98ca90b17450b16cfe59d235ce8c119b611b